### ISSUE TITLE
[Auth0] Not receiving a refresh token and error relating to wrong grantType

### ISSUE THREAD

**Issue by darthf1** [nuxt-community/auth-module#725]
I'm using `@nuxtjs/auth-next` with the `auth0` provider with the `PKCE Grant flow`. I'm having difficulties with setting up `auth0` and getting a `refreshToken`. 

My `nuxt.config.ts`: 
```js
  auth: {
    cookie: {
      options: {
        sameSite: 'strict',
        secure: true,
      },
    },
    localStorage: false,
    plugins: [{ src: '~/plugins/http', ssr: true }],
    redirect: {
      login: '/login/',
      logout: '/',
      callback: '/callback/',
      home: '/',
    },
    strategies: {
      auth0: {
        domain: 'xxx',
        clientId: 'xxx',
        audience: 'xxx',
        scope: ['openid', 'profile', 'email', 'offline_access'],
        responseType: 'token',
        grantType: 'authorization_code',
        codeChallengeMethod: 'S256',
      },
    },
  },
```

At first, I had the three grants enabled as displayed in the screenshot below.
![image](https://user-images.githubusercontent.com/17253332/84036619-b9c6e200-a99d-11ea-8b84-108e33ca7cd8.png)

The login was working properly and I was seeing some config related to the `PCKE`, however I was not receiving any refresh token (just the regulator token). 
![image](https://user-images.githubusercontent.com/17253332/84038056-b03e7980-a99f-11ea-8ba0-e4dc047b7df1.png)

I do have `offline access` is enabled in the auth0 settings:
![image](https://user-images.githubusercontent.com/17253332/84037103-6739f580-a99e-11ea-9d27-e7bdb275f20e.png)

Then I figured I wouldn't need the `implicit` grant flow, so I disabled that one. Now I'm getting the following error from `auth0`:

![image](https://user-images.githubusercontent.com/17253332/84037347-c435ab80-a99e-11ea-89d9-6cd5481df91d.png)

It tells me that the `implicit` grant is not allowed; which is true because I just disabled it. But I wouldn't expect the grantType `implicit` to be used, because I set it to `authorization_code`. Does this mean somethings is wrong with configuring the `grantType` and could this be the reason I'm not getting a refresh token?

**Comment by JoaoPedroAS51** [comment]
Hi @darthf1! I'm pretty sure that `responseType` should be `code`.

**Comment by darthf1** [comment]
Hi @JoaoPedroAS51, thanks, you're absolutely correct. The `unauthorized_client` error disappears. 

I managed to get it working, but there were still some hiccups:
- At first, I got redirected to the callback page and nothing was happening there. Fixed it by setting the token endpoint manually.
- Still stuck at the callback page, but now a `POST` request was made to the `/oauth/token` endpoint, but it returned a `401`.
- Had to change my application type from `Regular Web Application` to `Native` in Auth0 settings. Not really sure yet why I had to change this one.

- Do you want me to provide a PR to set the token endpoint it in the auth0 provider?
- Do you want me to provide a PR to add something to the docs regarding setting up PKCE with the auth0 provider?

**Comment by JoaoPedroAS51** [comment]
Hi @darthf1! I'm seeing here and `auth0` provider doesn't have support for this flow. I'll open a PR to add support to it :)

**Comment by JoaoPedroAS51** [comment]
> Do you want me to provide a PR to add something to the docs regarding setting up PKCE with the auth0 provider?

That would be great! 😃

**Comment by timhanlon** [comment]
> Had to change my application type from Regular Web Application to Native in Auth0 settings. Not really sure yet why I had to change this one.

@darthf1 thanks so much for this!

**Comment by darthf1** [comment]
> > Do you want me to provide a PR to set the token endpoint it in the auth0 provider?
> 
> > Do you want me to provide a PR to add something to the docs regarding setting up PKCE with the auth0 provider?
> 
> That would be great! Thanks! 😃

Provided in https://github.com/nuxt-community/auth-module/pull/730 and https://github.com/nuxt-community/auth-module/pull/746

**Comment by codeofsumit** [comment]
@darthf1 I tried to follow your example but I'm not getting a refresh token either.

This is my config:

```js
  auth: {
    plugins: [{ src: '~/plugins/preload', mode: 'client' }],
    rewriteRedirects: false,
    redirect: {
      home: '/portfolio',
      callback: '/logging-in/',
    },
    strategies: {
      local: false,
      auth0: {
        domain: process.env.VUE_APP_AUTHDOMAIN,
        clientId: process.env.VUE_APP_AUTHCLIENTID,
        audience: process.env.VUE_APP_AUTHAUDIENCE,
        scope: ['openid', 'profile', 'email', 'offline_access'],
        responseType: 'code',
        grantType: 'authorization_code',
        codeChallengeMethod: 'S256',
      },
    },
  },
```

Do you have an idea what I'm doing wrong?
Login/Logout works fine, but no refresh token in my cookies 🤔 

![image](https://user-images.githubusercontent.com/2399810/87461199-58360b00-c60e-11ea-9bd4-caf44ae3ac17.png)
![image](https://user-images.githubusercontent.com/2399810/87461236-6c7a0800-c60e-11ea-9cca-45fe177feaea.png)

I'd appreciate any help 👍

**Comment by darthf1** [comment]
Did you check your application type? Did you  enable offline access in the auth0 API settings?

**Comment by codeofsumit** [comment]
Hey @darthf1 

I have allowed offline access in the custom API Details whose identifier URL I pass in the "audience" parameter

![image](https://user-images.githubusercontent.com/2399810/87512513-708e4000-c677-11ea-995d-95f8ec9325b7.png)

Application Type is currently "Single Page Application". I tried with "Native Application" and "Regular Web Application" as well. What's the correct setting here?

![image](https://user-images.githubusercontent.com/2399810/87512660-b2b78180-c677-11ea-95ee-2cf754285212.png)

**Comment by darthf1** [comment]
> > Had to change my application type from Regular Web Application to Native in Auth0 settings. Not really sure yet why I had to change this one.
> 
> @darthf1 thanks so much for this!

Its just above your first post. This is what worked for me, dont know if its the 'correct' one.

**Comment by codeofsumit** [comment]
@darthf1 yea that's what I tried. Looking at the client cookies though, still shows refresh tokens as `false` 🤔 .
What else could it be?

**Comment by jawa-the-hutt** [comment]
wanted to drop a comment in this thread. 

tldr: Check that your Token Endpoint Authentication Method is set to `None` in the App Settings of the Auth0 GUI if you are getting the 401 Unauthorized on the code to access_token exchange.

I followed the new instructions that @darthf1 created in the PR's.  This included setting the App Type to `Native` but I was still getting the same symptom of the 401 Unauthorized when the POST was being made back to the `oauth/token` endpoint for the access_token exchange.  I was using the Default App that every Auth0 tenant provides automatically.

What I had to do was change the App Type back to `Regular Web App` and then change the Token Endpoint Authentication Method to `None`.  At that point I could change the App Type back to `Native` or `Single Page App ` and things worked either way.

When you create a brand new Auth0 app and if you set the App Type to `Native` or `Single Page App` during the creation, then Token Endpoint Authentication Method is already set to `None`.  However, if you choose `Regular Web App` then it is set to `Post`.  Since the Default App provided in the tenant starts as a `Regular Web App`, then this setting is wrong out of the gate.
